### PR TITLE
Refactor usb:ds, add 5.0.0+ support to usbComms.

### PR DESCRIPTION
**Pull request by SciresM**
What it says on the tin.

### CODE REVIEW

File: nx/source/runtime/devices/usb_comms.c [hunk 1]
```diff
@@ -25,69 +26,127 @@ static usbCommsInterface g_usbCommsInterfaces[TOTAL_INTERFACES];
...
+            
+            if (kernelAbove500()) {
+                u8 iManufacturer, iProduct, iSerialNumber;
+                static const u16 supported_langs[1] = {0x0409};
+                // Send language descriptor
+                rc = usbDsAddUsbLanguageStringDescriptor(NULL, supported_langs, sizeof(supported_langs)/sizeof(u16));
+                // Send manufacturer
+                if (R_SUCCEEDED(rc)) rc = usbDsAddUsbStringDescriptor(&iManufacturer, "Switchbrew");
+                // Send product
+                if (R_SUCCEEDED(rc)) rc = usbDsAddUsbStringDescriptor(&iProduct, "libnx USB comms");
```
Comment (SciresM) [inline nx/source/runtime/devices/usb_comms.c]: Should a different product string be sent here?

File: nx/source/runtime/devices/usb_comms.c [hunk 2]
```diff
@@ -25,69 +26,127 @@ static usbCommsInterface g_usbCommsInterfaces[TOTAL_INTERFACES];
...
+                    .idProduct = 0x3000,
+                    .bcdDevice = 0x0100,
+                    .iManufacturer = iManufacturer,
+                    .iProduct = iProduct,
+                    .iSerialNumber = iSerialNumber,
+                    .bNumConfigurations = 0x01
+                };
+                if (R_SUCCEEDED(rc)) rc = usbDsSetUsbDeviceDescriptor(UsbDeviceSpeed_Full, &device_descriptor);
+                
+                // High Speed is USB 2.0
```
Comment (yellows8) [inline nx/source/runtime/devices/usb_comms.c]: Add these comments to the UsbDeviceSpeed enum?

File: nx/source/runtime/devices/usb_comms.c [hunk 3]
```diff
@@ -25,69 +26,127 @@ static usbCommsInterface g_usbCommsInterfaces[TOTAL_INTERFACES];
...
+                    .bMaxPacketSize0 = 0x40,
+                    .idVendor = 0x057e,
+                    .idProduct = 0x3000,
+                    .bcdDevice = 0x0100,
+                    .iManufacturer = iManufacturer,
+                    .iProduct = iProduct,
+                    .iSerialNumber = iSerialNumber,
+                    .bNumConfigurations = 0x01
+                };
+                if (R_SUCCEEDED(rc)) rc = usbDsSetUsbDeviceDescriptor(UsbDeviceSpeed_Full, &device_descriptor);
```
Comment (yellows8) [inline nx/source/runtime/devices/usb_comms.c]: Add a comment for USB version here?

File: nx/source/runtime/devices/usb_comms.c [hunk 4]
```diff
@@ -25,69 +26,127 @@ static usbCommsInterface g_usbCommsInterfaces[TOTAL_INTERFACES];
...
+                if (R_SUCCEEDED(rc)) rc = usbDsSetUsbDeviceDescriptor(UsbDeviceSpeed_High, &device_descriptor);
+                
+                // Super Speed is USB 3.0
+                device_descriptor.bcdUSB = 0x0300;
+                // Upgrade packet size to 512
+                device_descriptor.bMaxPacketSize0 = 0x09;
+                if (R_SUCCEEDED(rc)) rc = usbDsSetUsbDeviceDescriptor(UsbDeviceSpeed_Super, &device_descriptor);
+                
+                // Define Binary Object Store
+                u8 bos[0x16] = {
```
Comment (yellows8) [inline nx/source/runtime/devices/usb_comms.c]: Is there no libusb struct for this?

File: nx/source/services/usb.c [hunk 5]
```diff
@@ -35,24 +35,21 @@ Result usbDsInitialize(UsbComplexId complexId, const UsbDsDeviceInfo* deviceinfo
...
         rc = _usbDsBindClientProcess(CUR_PROCESS_HANDLE);
 
     // GetStateChangeEvent
     if (R_SUCCEEDED(rc))
         rc = _usbDsGetEvent(&g_usbDsSrv, &g_usbDsStateChangeEvent, 3);
-
-    if (R_SUCCEEDED(rc) && deviceinfo && kernelAbove200()) {
-        rc = _usbDsSetVidPidBcd(deviceinfo);
-    }
+
```
Comment (yellows8) [inline nx/source/services/usb.c]: whitespace

File: nx/source/runtime/devices/usb_comms.c [hunk 6]
```diff
@@ -25,69 +26,127 @@ static usbCommsInterface g_usbCommsInterfaces[TOTAL_INTERFACES];
...
+    
+    if (g_usbCommsInitialized) {
+        rc = MAKERESULT(Module_Libnx, LibnxError_AlreadyInitialized); 
+    } else if (num_interfaces > TOTAL_INTERFACES) {
+        rc = MAKERESULT(Module_Libnx, LibnxError_OutOfMemory);
+    } else {
+        rc = usbDsInitialize();
+        
+        if (R_SUCCEEDED(rc)) {
+
```
Comment (yellows8) [inline nx/source/runtime/devices/usb_comms.c]: remove newline

File: nx/include/switch/services/usb.h [hunk 7]
```diff
@@ -153,38 +196,63 @@ enum usb_iso_usage_type {
...
-Service* usbDsGetServiceSession(void);
-Handle usbDsGetStateChangeEvent(void);
+/// Helpers
+Result usbDsWaitReady(u64 timeout);
+Result usbDsParseReportData(UsbDsReportData *reportdata, u32 urbId, u32 *requestedSize, u32 *transferredSize);
 
-Result usbDsGetState(u32 *out);
-Result usbDsGetDsInterface(UsbDsInterface** interface, struct usb_interface_descriptor* descriptor, const char *interface_name);
+/// IDsService
+// Do not provide API access to these functions, as they're handled by usbDsInitialize().
```
Comment (yellows8) [inline nx/include/switch/services/usb.h]: Remove these.

File: nx/source/services/usb.c [hunk 8]
```diff
@@ -265,7 +263,7 @@ static Result _usbDsGetEvent(Service* srv, Handle* handle_out, u64 cmd_id) {
         rc = resp->result;
 
         if (R_SUCCEEDED(rc)) {
-            *handle_out = r.Handles[0];
+            eventLoadRemote(event_out, r.Handles[0], false);
```
Comment (fincs) [inline nx/source/services/usb.c]: autoclear should probably be 'true' here, and instances of eventClear should be removed.

Comment (SciresM) [reply nx/source/services/usb.c]: I disagree, actually -- I think that autoclear should be false. The usb:ds API should provide a sufficiently low-level wrapping that custom interfaces can be written -- and so allowing manual clear should be the case.

Perhaps "eventSetAutoclear(Event *event, bool autoclear);" could be added, incidentally?

File: nx/source/runtime/devices/usb_comms.c (hunk 4 above)
Comment (SciresM) [reply nx/source/runtime/devices/usb_comms.c]: nope -- the relevant structs have nested variable length arrays...not really usable

File: nx/source/runtime/devices/usb_comms.c (hunk 3 above)
Comment (SciresM) [reply nx/source/runtime/devices/usb_comms.c]: Done

File: nx/source/runtime/devices/usb_comms.c (hunk 2 above)
Comment (SciresM) [reply nx/source/runtime/devices/usb_comms.c]: Done

File: nx/source/runtime/devices/usb_comms.c [hunk 9]
```diff
@@ -120,69 +174,167 @@ static void _usbCommsInterfaceExit(usbCommsInterface *interface)
...
+    Result rc = 0;
+    u32 ep_num = intf_ind + 1;
+    usbCommsInterface *interface = &g_usbCommsInterfaces[intf_ind];
+    
+    struct usb_interface_descriptor interface_descriptor = {
+        .bLength = USB_DT_INTERFACE_SIZE,
+        .bDescriptorType = USB_DT_INTERFACE,
+        .bInterfaceNumber = intf_ind,
+        .bNumEndpoints = 2,
+        .bInterfaceClass = 0xFF,
```
Comment (yellows8) [inline nx/source/runtime/devices/usb_comms.c]: USB_CLASS_VENDOR_SPEC, likewise below.

File: nx/source/services/usb.c (hunk 8 above)
Comment (fincs) [reply nx/source/services/usb.c]: The autoclear stuff parallels what Nintendo is doing in their IPC marshalling code. It seems like they decided some events will autoclear on the user's side, and some will never be cleared at all by the user. It sounds to me like this is one of the events that Nintendo mandated user-autoclear for.

Comment (SciresM) [reply nx/source/services/usb.c]: @fincs Correct, the relevant SystemEvents pass AutoClear = false for usb.

Comment (fincs) [reply nx/source/services/usb.c]: Fair enough. Seems like a weird design choice though.

File: nx/include/switch/services/usb.h [hunk 10]
```diff
@@ -35,13 +38,47 @@ struct usb_interface_descriptor {
...
     uint8_t  bInterfaceClass;
     uint8_t  bInterfaceSubClass;
     uint8_t  bInterfaceProtocol;
     uint8_t  iInterface; /// Ignored.
 };
 
+/// Imported from libusb, with some adjustments.
+struct usb_device_descriptor {
+    uint8_t  bLength;
+    uint8_t  bDescriptorType; /// Must match USB_DT_Device.
```
Comment (yellows8) [inline nx/include/switch/services/usb.h]: ///< + the below lines in enums.

File: nx/source/runtime/devices/usb_comms.c (hunk 9 above)
Comment (yellows8) [reply nx/source/runtime/devices/usb_comms.c]: Also let the user override these somehow.

File: nx/source/runtime/devices/usb_comms.c (hunk 1 above)
Comment (yellows8) [reply nx/source/runtime/devices/usb_comms.c]: Change it to what <5.0 uses.

File: nx/source/runtime/devices/usb_comms.c [hunk 11]
```diff
@@ -25,69 +26,135 @@ static usbCommsInterface g_usbCommsInterfaces[TOTAL_INTERFACES];
...
+    
+    if (R_SUCCEEDED(rc)) g_usbCommsInitialized = true;
 
     rwlockWriteUnlock(&g_usbCommsLock);
-
     return rc;
 }
 
-Result usbCommsInitialize(void)
+Result usbCommsInitialize(const UsbCommsInterfaceInfo *info)
```
Comment (yellows8) [inline nx/source/runtime/devices/usb_comms.c]: Can you change this to void and pass a hard-coded struct to usbCommsInitializeEx instead?

File: nx/include/switch/services/usb.h [hunk 12]
```diff
@@ -153,38 +196,60 @@ enum usb_iso_usage_type {
...
+/// IDsEndpoint
 void usbDsEndpoint_Close(UsbDsEndpoint* endpoint);
-Result usbDsEndpoint_PostBufferAsync(UsbDsEndpoint* endpoint, void* buffer, size_t size, u32 *urbId);
-Result usbDsEndpoint_GetReportData(UsbDsEndpoint* endpoint, UsbDsReportData *out);
+
+Result usbDsEndpoint_Cancel(UsbDsEndpoint* endpoint);
+Result usbDsEndpoint_PostBufferAsync(UsbDsEndpoint* endpoint, void* buffer, size_t size, u32* urbId);
+Result usbDsEndpoint_GetReportData(UsbDsEndpoint* endpoint, UsbDsReportData* out);
 Result usbDsEndpoint_StallCtrl(UsbDsEndpoint* endpoint);
+Result usbDsEndpoint_SetZlt(UsbDsEndpoint* endpoint, bool zlt);
```
Comment (yellows8) [inline nx/include/switch/services/usb.h]: Add comment for this?

**Comment by yellows8** [comment]
Broken on v2.x AFAICT :-(

**Comment by yellows8** [comment]
Works on v2.x now.

**Pull request merged**